File: jibjab/jjswift/Sources/jjswift/JJ/AST.swift
/// JibJab AST - Abstract Syntax Tree node definitions
/// Schema: ../../../common/ast.json

public protocol ASTNode {}

public struct Program: ASTNode {
    public let statements: [ASTNode]
}

public struct PrintStmt: ASTNode {
    public let expr: ASTNode
}

public struct InputExpr: ASTNode {
    public let prompt: ASTNode
}

public struct VarDecl: ASTNode {
    public let name: String
    public let value: ASTNode
}

public struct VarRef: ASTNode {
    public let name: String
}

public struct Literal: ASTNode {
    public let value: Any?
    public let numericType: NumericType?

    public init(value: Any?, numericType: NumericType? = nil) {
        self.value = value
        self.numericType = numericType
    }
}

public enum NumericType: String {
    case int = "Int"
    case int8 = "Int8"
    case int16 = "Int16"
    case int32 = "Int32"
    case int64 = "Int64"
    case uint = "UInt"
    case uint8 = "UInt8"
    case uint16 = "UInt16"
    case uint32 = "UInt32"
    case uint64 = "UInt64"
    case float = "Float"
    case double = "Double"
}

public struct BinaryOp: ASTNode {
    public let left: ASTNode
    public let op: String
    public let right: ASTNode
}

public struct UnaryOp: ASTNode {
    public let op: String
    public let operand: ASTNode
}

public struct LoopStmt: ASTNode {
    public let `var`: String
    public let start: ASTNode?
    public let end: ASTNode?
    public let collection: ASTNode?
    public let condition: ASTNode?
    public let body: [ASTNode]
}

public struct IfStmt: ASTNode {
    public let condition: ASTNode
    public let thenBody: [ASTNode]
    public let elseBody: [ASTNode]?
}

public struct FuncDef: ASTNode {
    public let name: String
    public let params: [String]
    public let body: [ASTNode]
}

public struct FuncCall: ASTNode {
    public let name: String
    public let args: [ASTNode]
}

public struct ReturnStmt: ASTNode {
    public let value: ASTNode
}

public struct ThrowStmt: ASTNode {
    public let value: ASTNode
}

public struct EnumDef: ASTNode {
    public let name: String
    public let cases: [String]
}

public struct ArrayLiteral: ASTNode {
    public let elements: [ASTNode]
}

public struct DictLiteral: ASTNode {
    public let pairs: [(key: ASTNode, value: ASTNode)]
}

public struct TupleLiteral: ASTNode {
    public let elements: [ASTNode]
}

public struct IndexAccess: ASTNode {
    public let array: ASTNode
    public let index: ASTNode
}

public enum StringInterpPart {
    case literal(String)
    case variable(String)
}

public struct StringInterpolation: ASTNode {
    public let parts: [StringInterpPart]
}

public struct TryStmt: ASTNode {
    public let tryBody: [ASTNode]
    public let oopsBody: [ASTNode]?
    public let oopsVar: String?
}

public struct CommentNode: ASTNode {
    public let text: String
}

File: jibjab/jjpy/jj/ast.py
"""
JibJab AST - Abstract Syntax Tree node definitions
"""

from typing import Any, List, Optional
from dataclasses import dataclass


@dataclass
class ASTNode:
    pass


@dataclass
class Program(ASTNode):
    statements: List[ASTNode]


@dataclass
class PrintStmt(ASTNode):
    expr: ASTNode


@dataclass
class InputExpr(ASTNode):
    prompt: ASTNode


@dataclass
class VarDecl(ASTNode):
    name: str
    value: ASTNode


@dataclass
class VarRef(ASTNode):
    name: str


@dataclass
class Literal(ASTNode):
    value: Any


@dataclass
class BinaryOp(ASTNode):
    left: ASTNode
    op: str
    right: ASTNode


@dataclass
class UnaryOp(ASTNode):
    op: str
    operand: ASTNode


@dataclass
class LoopStmt(ASTNode):
    var: str
    start: Optional[ASTNode]
    end: Optional[ASTNode]
    collection: Optional[ASTNode]
    condition: Optional[ASTNode]
    body: List[ASTNode]


@dataclass
class IfStmt(ASTNode):
    condition: ASTNode
    then_body: List[ASTNode]
    else_body: Optional[List[ASTNode]]


@dataclass
class FuncDef(ASTNode):
    name: str
    params: List[str]
    body: List[ASTNode]


@dataclass
class FuncCall(ASTNode):
    name: str
    args: List[ASTNode]


@dataclass
class ReturnStmt(ASTNode):
    value: ASTNode


@dataclass
class EnumDef(ASTNode):
    name: str
    cases: List[str]


@dataclass
class ArrayLiteral(ASTNode):
    elements: List[ASTNode]


@dataclass
class DictLiteral(ASTNode):
    pairs: List[tuple]  # List of (key: ASTNode, value: ASTNode) tuples


@dataclass
class TupleLiteral(ASTNode):
    elements: List[ASTNode]


@dataclass
class IndexAccess(ASTNode):
    array: ASTNode
    index: ASTNode
